docs/proposals/templates.md

```diff
@@ -0,0 +1,404 @@
+
+### Use cases for parameters within templates
+
+* Generate and share passwords between components (parameter value is provided to each component as an environment variable)
+* Allow for simple deployment-time customization of “app” configuration via environment values or api objects, e.g. memory 
+  tuning parameters to a MySQL image, Docker image registry prefix for image strings, pod resource requests and limits, default 
+  scale size.
+* Allow simple, declarative defaulting of parameter values AND expose them to end users in an approachable way - a parameter 
+  like “MySQL table space” can be parameterized in images as an env var - the template parameters declare the parameter, give 
+  it a friendly name, give it a reasonable default, and informs the user what tuning options available.
+* Customization of component names to avoid collisions and ensure matched labeling (eg replica selector value and pod label are 
+  user provided and in sync).
+* Customize cross-component references (eg user provides the name of a secret that already exists in their namespace, to use in 
+  a pod as a TLS cert).
+* Provide guidance to users for parameters such as default values, descriptions, and whether or not a particular parameter value 
+  is required or can be left blank.
+* Parameterize the replica count of a deployment or PetSet
+* Parameterize part of the labels and selector for a DaemonSet
+
+
+## Design Assumptions
+
+The goal for this proposal is a simple schema which addresses a few basic challenges:
+
+* Allow application authors to expose configuration knobs for application deployers, with suggested defaults and 
+descriptions of the purpose of each knob
+* Allow application deployers to easily customize exposed values like object names while maintaining referential integrity 
+  between dependent pieces
+* Support maintaining a library of templates within Kubernetes that can be accessed and instantiated by end users
+* Allow users to quickly and repeatedly deploy instances of well-defined application patterns produced by the community
+* Follow established Kubernetes API patterns by defining new template related APIs which consume+return first class Kubernetes 
+  API (and therefore json conformant) objects.
+
+We do not wish to invent a new turing complete templating language.  There are good options available 
+(e.g. https://github.com/mustache/mustache) for developers who want a completely flexible and powerful solution for creating 
+arbitrarily complex templates with parameters, and tooling can be built around such schemes.
+
+This desire for simplicity also intentionally excludes template composability/embedding as a supported use case.  
+
+Allowing templates to reference other templates presents versioning+consistency challenges along with making the template 
+no longer a self-contained portable object.  Scenarios necessitating multiple templates can be handled in one of several 
+alternate ways:
```

Comment:
It's quite straightforward to version templates, and to include template version information in references. See [this document](https://github.com/kubernetes/deployment-manager/blob/master/docs/templates/registry.md#template-versions) for an example of a simple but effective versioning scheme.

Beyond including template version information in references, a standardized reference format is essential, so that references can be unambiguously resolved against public registries. See [this document](https://github.com/kubernetes/deployment-manager/blob/master/docs/templates/registry.md#registry-based-template-references) for an example of Github based template references.

We should revise this document to require that templates be versioned and that templates residing in public registries be accessible through a standardized reference format.